### PR TITLE
add replay attack test

### PR DESCRIPTION
**Pull request by cyberliem**

### CODE REVIEW

File: evrclient/evrclient_test.go [hunk 1]
```diff
@@ -337,4 +346,127 @@ func TestChainID(t *testing.T) {
...
+	owner := common.HexToAddress("0x01")
+	provider := common.HexToAddress("0x02")
+	opts := types.CreateAccountOption{
+		OwnerAddress:    &owner,
+		ProviderAddress: &provider,
+	}
+	creationEnterpriseContractTx := types.NewContractCreation(uint64(1), big.NewInt(0), 1000000, big.NewInt(params.GasPriceConfig), data, opts)
+	creationEnterpriseContractTx, err = types.SignTx(creationEnterpriseContractTx, types.NewEIP155Signer(chainID), testKey2)
+	require.NoError(t, err)
+	creationEnterpriseContractTx, err = types.ProviderSignTx(creationEnterpriseContractTx, types.NewEIP155Signer(chainID), testKey)
```
Comment (ducquangkstn) [inline evrclient/evrclient_test.go:381]: Pls remove this. And add a TODO: creating contract should not be signed by provider

File: core/types/transaction_signing.go [hunk 2]
```diff
@@ -131,6 +134,8 @@ type Signer interface {
 	SignatureValues(tx *Transaction, sig []byte) (r, s, v *big.Int, err error)
 	// Hash returns the hash to be signed.
 	Hash(tx *Transaction) common.Hash
+	// HashWithSender returns the hash with sender address for provider to sign
+	HashWithSender(tx *Transaction) (common.Hash,error)
```
[the remark this answers is not in the document]
Comment (devnmhoang) [reply core/types/transaction_signing.go]: Removed

File: evrclient/evrclient_test.go [hunk 3]
```diff
@@ -337,4 +346,127 @@ func TestChainID(t *testing.T) {
...
+	require.NoError(t, err)
+
+	backend, _ := newTestBackend(t, types.Transactions{tx, txWithProvider, creationContractTx, creationEnterpriseContractTx})
```
[the remark this answers is not in the document]
Comment (devnmhoang) [reply evrclient/evrclient_test.go:384]: Added tests for creationEnterpriseContractTx

File: evrclient/evrclient_test.go (hunk 1 above)
Comment (devnmhoang) [reply evrclient/evrclient_test.go]: Updated

File: evrclient/evrclient_test.go [hunk 4]
```diff
@@ -378,8 +379,7 @@ func TestGetTransactionByHash(t *testing.T) {
 	creationEnterpriseContractTx := types.NewContractCreation(uint64(1), big.NewInt(0), 1000000, big.NewInt(params.GasPriceConfig), data, opts)
 	creationEnterpriseContractTx, err = types.SignTx(creationEnterpriseContractTx, types.NewEIP155Signer(chainID), testKey2)
 	require.NoError(t, err)
-	creationEnterpriseContractTx, err = types.ProviderSignTx(creationEnterpriseContractTx, types.NewEIP155Signer(chainID), testKey)
-	require.NoError(t, err)
+	//TODO: creating contract should not be signed by provider
```
Comment (cyberliem) [inline evrclient/evrclient_test.go:382]: Please verify that:
- The provider who sign the tx must be the provider of that contract.
- Following that logic, if the contract is not created yet, the tx should not be signed by provider (it hasn't got the list of provider yet).

Please add a test to invalidate create contract tx that has provider's signature

Comment (devnmhoang) [reply evrclient/evrclient_test.go:382]: After checking, I see that:
> The provider who sign the tx must be the provider of that contract.  

The Tx pool will validate this logic at https://github.com/Evrynetlabs/evrynet-node/blob/37062c406ca8e5fe01de3b302f5b01f1889ce3d9/core/tx_pool.go#L674 

> Following that logic, if the contract is not created yet, the tx should not be signed by provider (it hasn't got the list of provider yet).  

In this case, the ErrRedundantProvider error will be returned  https://github.com/Evrynetlabs/evrynet-node/blob/37062c406ca8e5fe01de3b302f5b01f1889ce3d9/core/tx_pool.go#L702

> Please add a test to invalidate create contract tx that has provider's signature

We already have a `TestCreateContractWithProviderSignature` test for this case

**Pull request merged**